src/test/java/wolox/training/controllers/UserControllerIntegrationTest.java

```diff
@@ -0,0 +1,154 @@
+  private BookRepository bookRepository;
+
+  @BeforeEach
+  void beforeEachTest(){
+    book = bookFactory.createTestBook();
+    jsonBook = bookFactory.getJsonBook(book);
+    user = userFactory.createTestUser();
+    user.setUserId(1);
+    jsonUser = userFactory.getJsonUser(user);
+  }
+
+  @Test
+  public void whenGetAllUsers_thenReturnUsersList() throws Exception {
+    List<User> users = new ArrayList<>();
+    users.add(user);
+
+    given(userRepository.findAll()).willReturn(users);
+
+    mvc.perform(get("/api/users")
+        .contentType(MediaType.APPLICATION_JSON))
+        .andExpect(jsonPath("$", hasSize(1)))
+        .andExpect(jsonPath("$[0].username", is(user.getUsername())));
+  }
+
+  @Test
+  public void whenFindUserByIdAndUserDoesNotExists_thenReturnNotFoundException() throws Exception {
+    given(userRepository.findById(user.getUserId())).willReturn(Optional.empty());
+
+    mvc.perform(get("/api/users/".concat(Long.toString(user.getUserId())))
+    .contentType(MediaType.APPLICATION_JSON))
+        .andExpect(status().isNotFound());
+  }
+
+  @Test
+  public void whenFindUserByUsername_thenReturnUser() throws Exception {
+    given(userRepository.findByUsername(user.getUsername())).willReturn(Optional.ofNullable(user));
+
+    mvc.perform(get("/api/users/username/".concat(user.getUsername()))
+        .contentType(MediaType.APPLICATION_JSON))
+        .andExpect(status().isOk())
+        .andExpect(jsonPath("$.username", is(user.getUsername())));
+  }
+
+  @Test
+  public void whenCreateUser_thenReturnUser() throws Exception {
+    given(userRepository.save(any(User.class))).willReturn(user);
+
+    mvc.perform(post("/api/users/")
+    .contentType(MediaType.APPLICATION_JSON)
+    .content(jsonUser))
+        .andExpect(status().isCreated())
+        .andExpect(jsonPath("$.username", is(user.getUsername())));
+  }
+
+  @Test
+  public void  whenDeleteUserAndDoesNotExists_thenReturnNotFoundException() throws Exception {
+    given(userRepository.findById(user.getUserId())).willReturn(Optional.empty());
+
+    mvc.perform(delete("/api/users/".concat(Long.toString(user.getUserId())))
+        .contentType(MediaType.APPLICATION_JSON))
+        .andExpect(status().isNotFound());
+  }
+
+  @Test
+  public void whenUpdateUserAndIdMismatch_thenReturnIdMismatchException() throws Exception {
+    mvc.perform(put("/api/users/2")
+    .contentType(MediaType.APPLICATION_JSON)
+    .content(jsonUser))
+        .andExpect(status().isPreconditionFailed());
+  }
+
+  @Test
+  public void whenAddBookToUser_thenReturnUserWithBook() throws Exception {
+    book.setId(1);
+    jsonBook = bookFactory.getJsonBook(book);
+    given(userRepository.findById(user.getUserId())).willReturn(Optional.ofNullable(user));
+    given(bookController.findOne(book.getId())).willReturn(book);
+    given(userRepository.save(any(User.class))).willReturn(user);
+
+    mvc.perform(put("/api/users/".concat(Long.toString(user.getUserId())).concat("/books/add"))
```

Comment:
What about defining the ruote in a constant like this:
`private static final String ADD_BOOK_PATH = "/api/users/%1$s/books/add"`
and then do:
`var path = String.format(ADD_BOOK_PATH, user.getUserId());`